mongoengine/base/document.py

```diff
@@ -746,14 +735,15 @@ def merge_index_specs(index_specs, indices):
             if not indices:
                 return index_specs
 
-            spec_fields = [v['fields']
-                           for k, v in enumerate(index_specs)]
+            spec_fields = [idx['fields'] for idx in index_specs]
```

Comment:
I haven't dug very deeply, but `idx['fields']` can definitely be unhashable as-is. For example, see https://github.com/MongoEngine/mongoengine/blob/25e0f12976635c0c6fc89ea68b2eacfdda2fce55/mongoengine/base/document.py#L923 where it's a list.

That being said, perhaps it would be easy to turn these potential lists into hashable tuples. After all, an index you define in the code shouldn't be mutated during runtime...